### PR TITLE
Avascience

### PR DESCRIPTION
**Pull request by avavita**
"It is a wise man that plays the fool."

### CODE REVIEW

File: data_science/spec/split_test_spec.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  describe '#confidence' do
+    it 'finds min/max predictions for a given confidence level' do
+      expect(test.confidence[0]).to eq([0.5300927453027154, 0.6155383226584495])
+    end
+  end
+
+  describe '#chi_square' do
+    it 'returns the p values for a chi_square test' do
+      expect(test.chi_square).to eq(5.1514348342607263e-14)
+    end
```
Comment (zsyed1) [inline data_science/spec/split_test_spec.rb]: :smile: 5e-14 is awfully close to zero!
Don't sweat it though.

Comment (zsyed1) [inline data_science/spec/split_test_spec.rb]: Did you generate the data randomly? (`test_data.json`)
That would make the chi-square close to zero. If instead you made the proportions of successes high like 75% for A and 50% of B you would get a better value (for testing). This is not a coding concern though so don't worry about it.

File: data_science/split_test.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      values[cohort.name] =
+      {
+        converted: cohort.successes,
+        unconverted: cohort.failures
+      }
+    end
+    tester = ABAnalyzer::ABTest.new values
+    tester.chisquare_p
+  end
+end
```
Comment (zsyed1) [inline data_science/split_test.rb]: Very clean! This should be closed @djkotowski

File: data_science/ab_runner.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+require 'pp'
+require_relative 'split_test'
+require_relative 'cohort_converter'
+
+abort('Please include one json file as an argument.') unless ARGV.count == 1
+
+cohorts = CohortConverter.new(ARGV[0]).cohorts
+test = SplitTest.new(cohorts)
+
+successes = cohorts.each_with_object(Hash.new) { |cohort, memo| memo[cohort.name] = cohort.successes }
```
Comment (djkotowski) [inline data_science/ab_runner.rb]: `Hash.new` can be `{}`

File: data_science/spec/cohort_converter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  describe '#cohorts' do
+    it 'returns a collection of cohorts.' do
+      data.cohorts.each do |cohort|
+        expect(cohort).to be_a(Cohort)
+      end
+    end
+  end
+end
```
Comment (djkotowski) [inline data_science/spec/cohort_converter_spec.rb]: What about expected result if data is invalid?

File: data_science/spec/split_test_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
+require 'spec_helper'
+
+describe SplitTest do
+  let(:converter) { CohortConverter.new('spec/test_data.json') }
+  let(:test) { described_class.new(converter.cohorts) }
+
+  describe '#new' do
+    it 'takes a collection of cohorts and returns a SplitTest object' do
+      expect(test).instance_of? described_class
```
Comment (djkotowski) [inline data_science/spec/split_test_spec.rb]: This doesn't read quite as clearly as
```ruby
expect(test).to be_a(described_class)
```

File: data_science/spec/split_test_spec.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    end
+  end
+
+  describe '#cohorts' do
+    it 'returns a collection of cohorts' do
+      expect(test).instance_of? described_class
```
Comment (djkotowski) [inline data_science/spec/split_test_spec.rb]: Copy/paste error here.  The test doesn't match its description.

File: data_science/spec/split_test_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    end
+  end
+
+  describe '#conversion_rates'
+    it 'returns a collection of conversion rates, one for each cohort' do
+      expect(test.conversion_rates[0]).to eq(0.5728155339805825)
```
Comment (djkotowski) [inline data_science/spec/split_test_spec.rb]: The `be_within` expectation can help you avoid having to be precise in these cases.

File: data_science/split_test.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  def chi_square
+    values = {}
+    @cohorts.each do |cohort|
+      values[cohort.name] =
+      {
+        converted: cohort.successes,
+        unconverted: cohort.failures
+      }
+    end
+    tester = ABAnalyzer::ABTest.new values
```
Comment (djkotowski) [inline data_science/split_test.rb]: Use parentheses to wrap initializer args.

File: data_science/split_test.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+  def confidence(level_of_confidence = 0.95)
+    @cohorts.map { |cohort| ABAnalyzer.confidence_interval(cohort.successes, cohort.attempts, level_of_confidence) }
+  end
+
+  def chi_square
+    values = {}
+    @cohorts.each do |cohort|
+      values[cohort.name] =
+      {
```
Comment (djkotowski) [inline data_science/split_test.rb]: Indent to make clearer.  You could also use a different alignment to save lines:

```ruby
values[cohort.name] = { converted: cohort.successes
                        unconverted: cohort.failures }

**Comment by djkotowski** [comment]
Well done!  Take a look at the comments, but no need for resubmission.  You're done.

File: data_science/spec/cohort_converter_spec.rb (hunk 4 above)
Comment (avavita) [inline data_science/spec/cohort_converter_spec.rb]: We're planning on testing this via customer satisfaction surveys.

Comment (djkotowski) [inline data_science/spec/cohort_converter_spec.rb]: :laughing:

File: data_science/ab_runner.rb (hunk 3 above)
Comment (avavita) [inline data_science/ab_runner.rb]: I guess.

**Pull request closed without merging**